feat: make SpaceBindingRequest.Spec.MasterUserRecord immutable

mfrancisc opened the pull request:
This PR introduces a new validating webhook for SpaceBindingRequests. Specifically the webhook makes sure that once an SBR is created the MasterUserRecord field cannot be changed anymore. Changing the MUR field in SBR will generate an discrepancy with the name of SpaceBinding that will still contain the old MUR.

Jira: https://issues.redhat.com/browse/ASC-426

Paired with: https://github.com/codeready-toolchain/toolchain-e2e/pull/792

mfrancisc commented:
/retest

updated e2e PR

MatousJobanek commented on pkg/webhook/validatingwebhook/validate_rolebinding_request_test.go:
```diff
@@ -2,24 +2,23 @@ package validatingwebhook
 )
 
-func TestHandleValidateRolebBndingAdmissionRequestBlocked(t *testing.T) {
+func TestHandleValidateRolebBindingAdmissionRequestBlocked(t *testing.T) {
```
is the extra `b` in `Roleb` expected? 
it's the same also in other names of the test functions

mfrancisc replied:
While trying to fix a typo I've introduced a new one 🤦‍♂️ . Thanks it should be fixed in https://github.com/codeready-toolchain/member-operator/pull/470/commits/9ddde444e5b80865e6c6793321a2dfcfff9f15b0

MatousJobanek commented on pkg/webhook/validatingwebhook/validate_spacebindingrequest.go:
```diff
@@ -0,0 +1,85 @@
+		}
+		// there was an issue while trying to GET SBR
+		log.Error(err, "unable to check if spacebindingrequest already exists", "SpaceBindingRequest.Name", newSBR.GetName(), "SpaceBindingRequest.Namespace", newSBR.GetNamespace())
+		return denyAdmissionRequest(admReview, errs.Wrapf(err, "unable to check if spacebindingrequest already exists. SpaceBindingRequest.Name: %v", newSBR.GetName()))
```
maybe something more generic would be better
```suggestion
		return denyAdmissionRequest(admReview, errs.Wrapf(err, "unable to validate the SpaceBindingRequest. SpaceBindingRequest.Name: %v", newSBR.GetName()))
```

MatousJobanek commented on pkg/webhook/validatingwebhook/validate_spacebindingrequest.go:
```diff
@@ -0,0 +1,85 @@
+	if existingSBR.Spec.MasterUserRecord != newSBR.Spec.MasterUserRecord {
+		// MUR name field is immutable since the SpaceBinding name contains the MUR name,
+		// changing it, then it wouldn't match anymore.
+		return denyAdmissionRequest(admReview, errs.New("SpaceBindingRequest.MasterUserRecord field cannot be updated. Consider deleting and recreating the SpaceBindingRequest resource"))
```
```suggestion
		return denyAdmissionRequest(admReview, errs.New("SpaceBindingRequest.MasterUserRecord field cannot be changed. Consider deleting and creating a new SpaceBindingRequest resource"))
```

MatousJobanek commented on pkg/webhook/validatingwebhook/validate_spacebindingrequest_test.go:
```diff
@@ -0,0 +1,251 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels: map[string]string{
+				toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
+			},
```
why do we need this label set?

mfrancisc replied:
for the sake of the test we don't , was a copy paste left over I guess. Removed in https://github.com/codeready-toolchain/member-operator/pull/470/commits/9ddde444e5b80865e6c6793321a2dfcfff9f15b0

MatousJobanek reviewed (commented):
Looks good overall :+1: Just a few comments

mfrancisc commented on deploy/webhook/member-operator-webhook.yaml:
```diff
@@ -185,6 +193,30 @@ objects:
+          port: 443
+      matchPolicy: Equivalent
+      rules:
+        - operations: ["CREATE", "UPDATE"]
```
I remember adding PATCH initially but was failing, since apparently there is not PATCH operations 🤷‍♂️ :https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules

> operations lists one or more operations to match. Can be "CREATE", "UPDATE", "DELETE", "CONNECT", or "*" to match all.

MatousJobanek replied:
ah, OK. My bad - I automatically expected that Patch would be one of the operations as well :-)
Thanks for the link and sorry for confusion :+1:

mfrancisc replied:
No worries, I expected the same and I'm still a bit confused of why are those the "operations". What is connect for example 😕

mfrancisc commented on make/go.mk:
```diff
@@ -34,9 +34,9 @@ vendor:
-	@$(GO_BINDATA) -pkg userspodswebhook -o ./pkg/webhook/deploy/userspodswebhook/template_assets.go -nocompress -prefix deploy/webhook deploy/webhook
+	@echo "generating webhooks template data..."
+	@rm ./pkg/webhook/deploy/templates/template_assets.go 2>/dev/null || true
+	@$(GO_BINDATA) -pkg templates -o ./pkg/webhook/deploy/templates/template_assets.go -nocompress -prefix deploy/webhook deploy/webhook
```
I've renamed it to `webhooks` plz check : https://github.com/codeready-toolchain/member-operator/pull/470/commits/9ddde444e5b80865e6c6793321a2dfcfff9f15b0

mfrancisc commented:
@MatousJobanek thanks for your review. I've addressed your comments, when you have time PTAL. I've also fixed/updated some unit tests which apparently I've missed initially.

mfrancisc commented:
/retest

Flaky createSpace? https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/codeready-toolchain_member-operator/470/pull-ci-codeready-toolchain-member-operator-master-e2e/1699788328947159040

I haven't saw this in a while.

mfrancisc commented:
/retest

updated e2e PR

mfrancisc commented:
/retest

infra

mfrancisc commented:
/retest

infra

mfrancisc commented:
/retest 

infra

mfrancisc commented:
/retest

infra

MatousJobanek reviewed (approved):
Thanks :+1:

mfrancisc commented:
/retest

infra

mfrancisc commented:
/retest

infra
